design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,169 @@
+# Adding and removing Ceph osds via Ceph orchestrator interface
+
+## Goals
+
+1. Give Ceph's Rook Orchestrator Plugin the ability to add and remove OSDs to/from a Rook-Ceph
+   cluster running in Kubernetes.
+2. Users of the Ceph Orchestrator CLI and/or Ceph Dashboard should be able to
+   create OSDs on nodes with as much complexity as Ceph is able.
+3. Do not introduce unnecessary complexity into Rook or into Rook's CephCluster CRD.
+
+## Proposal
+
+### Removing OSDs
+Removing OSDs is the simplest place to start with this design. When removing an OSD or OSDs via the
+Rook Orchestrator Module, the Ceph mgr running in Kubernetes will run a Kubernetes Job that will run
+the required `ceph-volume` command necessary to remove the OSD(s) from a node as desired by the Ceph
+user. The Orchestrator Module will also perform whatever Ceph operations are necessary to remove the
+OSD as would be expected by the user. This will replace the manual, imperative steps currently
+required by Rook-Ceph users to remove OSDs from the storage cluster.
+
+#### OSD removal in the CephCluster resource
+The CephCluster CRD allows adding disks to nodes in several places including disks on all nodes in
+the Kubernetes cluster, to a specific node, or via disk pattern matching (in two different ways). To
+avoid the complexity of having to manage the logic for determining how to remove a node from the
+CephCluster resource, OSD removal will **not** alter the CephCluster CRD. While doing so would be
+convenient for users, there are a lot of opportunities in development to miss corner cases or
+introduce logic bugs. This may mean that Rook could re-add the disk after the disk is removed via
+the Ceph Orchestrator interface if the CephCluster CRD is configured to do so, and the user should
+be made aware of this possibility via documentation.
+
+### Adding OSDs
+When adding an OSD or OSDs via the Rook Orchestrator Module, the Ceph mgr running in Kubernetes will
+run a Kubernetes Job that will run the required `ceph-volume` command necessary to provision the
```

Comment:
Instead of running arbitrary c-v commands, what if we triggered the osd prepare job to do the provisioning step? This seems like a more natural step to integrate with the orchestration instead of doing something completely independent with c-v. 

That said, my view is still that we should use the CR to add all OSDs until we prove why we can't use it. The complexity of coordinating two different ways to add OSDs just isn't manageable in my head.

Reply:
Agree with @Travis. I think that addition/removal of OSd's should start with the modification of the cluster CRD, and let the reconcile loop to do the job.
To have two ways of adding/removing OSd's will make things more complicated and event produce unexpected results sometimes.

Reply:
To do this, we will have to either do option  1 or 2 lower down in the doc.